In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 

In [2]:
df = pd.read_csv("data_with_coord.csv")

/home/smail/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (0,5) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
pd.set_option('display.max_columns', 25)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1190683 entries, 0 to 1190682
Data columns (total 25 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   IncidentNumber            1190683 non-null  object 
 1   TimeMobilised             1190683 non-null  object 
 2   TimeMobile                1190683 non-null  object 
 3   TimeArrived               1190683 non-null  object 
 4   TimeLeft                  1148452 non-null  object 
 5   TimeReturned              974961 non-null   object 
 6   DeployedFromStation_Name  1190683 non-null  object 
 7   DateOfCall                1190683 non-null  object 
 8   CalYear                   1190683 non-null  int64  
 9   TimeOfCall                1190683 non-null  object 
 10  HourOfCall                1190683 non-null  int64  
 11  IncidentGroup             1190683 non-null  object 
 12  SpecialServiceType        275800 non-null   object 
 13  Easting_rounded           1

In [5]:
df.head()

IncidentNumber        TimeMobilised           TimeMobile  \
0      235138081  01/01/2009 00:02:27  01/01/2009 00:06:40   
1           2091  01/01/2009 00:04:09  01/01/2009 00:06:40   
2           3091  01/01/2009 00:04:57  01/01/2009 00:06:45   
3           5091  01/01/2009 00:06:04  01/01/2009 00:08:26   
4           5091  01/01/2009 00:06:04  01/01/2009 00:07:58   

           TimeArrived             TimeLeft         TimeReturned  \
0  01/01/2009 00:08:09  01/01/2009 00:12:17  01/01/2009 00:12:17   
1  01/01/2009 00:09:17  01/01/2009 00:16:33  01/01/2009 00:17:48   
2  01/01/2009 00:08:27  01/01/2009 00:20:38  01/01/2009 00:20:38   
3  01/01/2009 00:10:14                  NaN  01/01/2009 00:18:53   
4  01/01/2009 00:09:57  01/01/2009 00:11:58  01/01/2009 00:22:29   

  DeployedFromStation_Name  DateOfCall  CalYear TimeOfCall  HourOfCall  \
0                  Clapham  01-janv-09     2009   00:00:37           0   
1                 Edmonton  01-janv-09     2009   00:03:00           0   
2               Hillingdon  01-janv-09     2009   00:04:27           0   
3                 Holloway  01-janv-09     2009   00:05:39           0   
4                 Holloway  01-janv-09     2009   00:05:39           0   

     IncidentGroup SpecialServiceType  Easting_rounded  Northing_rounded  \
0  Special Service                RTC           528650            176850   
1             Fire                NaN           533750            194450   
2             Fire                NaN           507750            182850   
3             Fire                NaN           531050            185350   
4             Fire                NaN           531050            185350   

  color    d_long      d_lat         x_utm         y_utm    o_long      o_lat  \
0  blue -0.148916  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400   
1   red -0.068836  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217   
2   red -0.447902  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257   
3   red -0.111220  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766   
4   red -0.111220  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766   

        o_x_utm       o_y_utm  dist_euclidian  
0 -15701.725496  6.704034e+06     2085.533280  
1  -7694.403204  6.733000e+06     1032.257024  
2 -50283.125311  6.715810e+06      687.205480  
3 -12964.379217  6.721272e+06     1872.413548  
4 -12964.379217  6.721272e+06     1872.413548

# Transformons les colonnes "Time" en type datetime, ce qui va faciliter la sélection, (ça prend un peu de temps):

In [6]:
df['TimeMobilised'] = pd.to_datetime(df['TimeMobilised'])

In [7]:
df['TimeMobile'] = pd.to_datetime(df['TimeMobile'])

In [8]:
df['TimeArrived'] = pd.to_datetime(df['TimeArrived'])

In [9]:
df['TimeLeft'] = pd.to_datetime(df['TimeLeft'])

In [10]:
df['TimeReturned'] = pd.to_datetime(df['TimeReturned'])

In [11]:
from datetime import datetime

# Un aperçu au hasard de la facilité de sélectionner les colonnes grâce aux attributs propres au type datetime, ce qui va grandement faciliter les dataviz: 

In [12]:
df.loc[(df['TimeMobilised'].dt.year == 2012) & (df['TimeMobilised'].dt.month == 1) 
        & (df['TimeMobilised'].dt.hour == 14) & (df['TimeArrived'].dt.minute == 15)]

IncidentNumber       TimeMobilised          TimeMobile  \
490591         319121 2012-01-01 14:11:57 2012-01-01 14:13:25   
490592         318121 2012-01-01 14:12:18 2012-01-01 14:13:54   
501957       11791121 2012-01-31 14:00:27 2012-01-31 14:13:51   
502063       11923121 2012-01-31 14:06:25 2012-01-31 14:07:54   
515355       28472121 2012-01-03 14:06:15 2012-01-03 14:06:52   
515357       28476121 2012-01-03 14:11:19 2012-01-03 14:12:27   
563984       24205122 2012-01-07 14:09:51 2012-01-07 14:10:44   
563986       24205122 2012-01-07 14:09:51 2012-01-07 14:11:02   
614639      134486121 2012-01-11 14:13:38 2012-01-11 14:13:56   
626669      149238121 2012-01-12 14:09:59 2012-01-12 14:10:45   
626670      149238121 2012-01-12 14:09:59 2012-01-12 14:11:24   

               TimeArrived            TimeLeft        TimeReturned  \
490591 2012-01-01 14:15:24 2012-01-01 14:42:51 2012-01-01 14:45:56   
490592 2012-01-01 14:15:51 2012-01-01 14:20:42 2012-01-01 14:23:46   
501957 2012-01-31 15:15:14 2012-01-31 19:34:00 2012-01-31 21:06:11   
502063 2012-01-31 14:15:31 2012-01-31 14:36:04 2012-01-31 14:47:37   
515355 2012-01-03 14:15:28 2012-01-03 14:15:43 2012-01-03 14:19:19   
515357 2012-01-03 14:15:16 2012-01-03 14:33:06 2012-01-03 14:35:17   
563984 2012-01-07 14:15:13 2012-01-07 14:20:09 2012-01-07 14:20:09   
563986 2012-01-07 14:15:26 2012-01-07 14:23:00 2012-01-07 14:31:26   
614639 2012-01-11 14:15:50 2012-01-11 14:26:49 2012-01-11 14:36:11   
626669 2012-01-12 14:15:49 2012-01-12 14:23:25 2012-01-12 14:31:44   
626670 2012-01-12 14:15:47 2012-01-12 14:22:38 2012-01-12 14:22:38   

       DeployedFromStation_Name    DateOfCall  CalYear TimeOfCall  HourOfCall  \
490591                 Deptford    01-janv-12     2012   14:10:53          14   
490592                New Cross    01-janv-12     2012   14:10:41          14   
501957                Plumstead    31-janv-12     2012   09:35:31           9   
502063                Willesden    31-janv-12     2012   14:05:40          14   
515355              Whitechapel    01-mars-12     2012   14:06:04          14   
515357                   Bexley    01-mars-12     2012   14:10:43          14   
563984                 Finchley    01-juil-12     2012   14:09:33          14   
563986                   Hendon    01-juil-12     2012   14:09:33          14   
614639                  Brixton     01-nov-12     2012   14:11:01          14   
626669                Wimbledon  01-déc-12     2012   14:09:34          14   
626670                  Tooting  01-déc-12     2012   14:09:34          14   

          IncidentGroup SpecialServiceType  Easting_rounded  Northing_rounded  \
490591  Special Service           Flooding           535850            178550   
490592  Special Service       Lift Release           534750            176750   
501957             Fire                NaN           523350            180950   
502063      False Alarm                NaN           523250            180750   
515355             Fire                NaN           533550            179950   
515357      False Alarm                NaN           549250            175250   
563984      False Alarm                NaN           523950            188350   
563986      False Alarm                NaN           523950            188350   
614639      False Alarm                NaN           532650            176150   
626669      False Alarm                NaN           526350            172050   
626670      False Alarm                NaN           526350            172050   

       color    d_long      d_lat         x_utm         y_utm    o_long  \
490591  blue -0.044651  51.489587  -4970.478716  6.708357e+06 -0.034350   
490592  blue -0.061171  51.473674  -6809.549062  6.705513e+06 -0.047339   
501957   red -0.223760  51.514019 -24908.896014  6.712726e+06  0.013175   
502063  cyan -0.225271  51.512244 -25077.020728  6.712409e+06 -0.236838   
515355   red -0.077229  51.502716  -8597.117445  6.710705e+06 -0

# A mon avis, on n'aura plus besoin des colonnes 'DateOfCall' et 'CallYear', vu que tout est repris dans les premières colonnes déjà transformées. 

In [13]:
df.drop(['DateOfCall','CalYear'], axis = 1, inplace = True)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1190683 entries, 0 to 1190682
Data columns (total 23 columns):
 #   Column                    Non-Null Count    Dtype         
---  ------                    --------------    -----         
 0   IncidentNumber            1190683 non-null  object        
 1   TimeMobilised             1190683 non-null  datetime64[ns]
 2   TimeMobile                1190683 non-null  datetime64[ns]
 3   TimeArrived               1190683 non-null  datetime64[ns]
 4   TimeLeft                  1148452 non-null  datetime64[ns]
 5   TimeReturned              974961 non-null   datetime64[ns]
 6   DeployedFromStation_Name  1190683 non-null  object        
 7   TimeOfCall                1190683 non-null  object        
 8   HourOfCall                1190683 non-null  int64         
 9   IncidentGroup             1190683 non-null  object        
 10  SpecialServiceType        275800 non-null   object        
 11  Easting_rounded           1190683 non-null  int64 

# Je pense qu'il est judicieux de transformer aussi la colonne 'TimeOfCall' en type datetime, puis de la placer de préférence en index n°1 du dataframe, pour plus de lisibilité. Elle nous permettra surtout de calculer le temps de réponse plus loin.

In [15]:
df['TimeOfCall'] = pd.to_datetime(df['TimeOfCall'])

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1190683 entries, 0 to 1190682
Data columns (total 23 columns):
 #   Column                    Non-Null Count    Dtype         
---  ------                    --------------    -----         
 0   IncidentNumber            1190683 non-null  object        
 1   TimeMobilised             1190683 non-null  datetime64[ns]
 2   TimeMobile                1190683 non-null  datetime64[ns]
 3   TimeArrived               1190683 non-null  datetime64[ns]
 4   TimeLeft                  1148452 non-null  datetime64[ns]
 5   TimeReturned              974961 non-null   datetime64[ns]
 6   DeployedFromStation_Name  1190683 non-null  object        
 7   TimeOfCall                1190683 non-null  datetime64[ns]
 8   HourOfCall                1190683 non-null  int64         
 9   IncidentGroup             1190683 non-null  object        
 10  SpecialServiceType        275800 non-null   object        
 11  Easting_rounded           1190683 non-null  int64 

# Changeons l'ordre des colonnes: 

In [17]:
df.columns

Index(['IncidentNumber', 'TimeMobilised', 'TimeMobile', 'TimeArrived',
       'TimeLeft', 'TimeReturned', 'DeployedFromStation_Name', 'TimeOfCall',
       'HourOfCall', 'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
       'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
       'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian'],
      dtype='object')

In [18]:
new_col_order = ['IncidentNumber','TimeOfCall','TimeMobilised', 'TimeMobile', 'TimeArrived',
                   'TimeLeft', 'TimeReturned', 'DeployedFromStation_Name','HourOfCall', 
                 'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
                   'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
                   'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian']

In [19]:
df = df.reindex(columns = new_col_order)

In [20]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

          TimeArrived            TimeLeft        TimeReturned  \
0 2009-01-01 00:08:09 2009-01-01 00:12:17 2009-01-01 00:12:17   
1 2009-01-01 00:09:17 2009-01-01 00:16:33 2009-01-01 00:17:48   
2 2009-01-01 00:08:27 2009-01-01 00:20:38 2009-01-01 00:20:38   
3 2009-01-01 00:10:14                 NaT 2009-01-01 00:18:53   
4 2009-01-01 00:09:57 2009-01-01 00:11:58 2009-01-01 00:22:29   

  DeployedFromStation_Name  HourOfCall    IncidentGroup SpecialServiceType  \
0                  Clapham           0  Special Service                RTC   
1                 Edmonton           0             Fire                NaN   
2               Hillingdon           0             Fire                NaN   
3                 Holloway           0             Fire                NaN   
4                 Holloway           0             Fire                NaN   

   Easting_rounded  Northing_rounded color    d_long      d_lat         x_utm  \
0           528650            176850  blue -0.148916  51.475992 -16577.246612   
1           533750            194450   red -0.068836  51.632970  -7662.804053   
2           507750            182850   red -0.447902  51.534284 -49860.215886   
3           531050            185350   red -0.111220  51.551829 -12380.899219   
4           531050            185350   red -0.111220  51.551829 -12380.899219   

          y_utm    o_long      o_lat       o_x_utm       o_y_utm  \
0  6.705927e+06 -0.141051  51.465400 -15701.725496  6.704034e+06   
1  6.734032e+06 -0.069120  51.627217  -7694.403204  6.733000e+06   
2  6.716352e+06 -0.451701  51.531257 -50283.125311  6.715810e+06   
3  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   
4  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   

   dist_euclidian  
0     2085.533280  
1     1032.257024  
2      687.205480  
3     1872.413548  
4     1872.413548

# Une boucle for pour créer la liste des temps de réponse en secondes:

In [21]:
liste = []
for i in range(0, len(df)):
    liste.append((df['TimeArrived'][i] - df['TimeOfCall'][i]).seconds)

# Puis création d'un dataframe contenant la liste en question:

In [22]:
ResponseTime = pd.DataFrame({'ResponseTime': liste})

In [23]:
ResponseTime.head()

ResponseTime
0           452
1           377
2           240
3           275
4           258

# Puis jonction de df et ResponseTime:

In [24]:
df = df.join(ResponseTime)

# On replace la colonne 'ResponseTime' à l'index n°5 pour plus de lisibilité:

In [25]:
df.columns

Index(['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile',
       'TimeArrived', 'TimeLeft', 'TimeReturned', 'DeployedFromStation_Name',
       'HourOfCall', 'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
       'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
       'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian',
       'ResponseTime'],
      dtype='object')

In [26]:
new_col_order = ['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile',
       'TimeArrived','ResponseTime', 'TimeLeft', 'TimeReturned', 'DeployedFromStation_Name',
       'HourOfCall', 'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
       'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
       'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian']

In [27]:
df = df.reindex(columns = new_col_order)

In [28]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

          TimeArrived  ResponseTime            TimeLeft        TimeReturned  \
0 2009-01-01 00:08:09           452 2009-01-01 00:12:17 2009-01-01 00:12:17   
1 2009-01-01 00:09:17           377 2009-01-01 00:16:33 2009-01-01 00:17:48   
2 2009-01-01 00:08:27           240 2009-01-01 00:20:38 2009-01-01 00:20:38   
3 2009-01-01 00:10:14           275                 NaT 2009-01-01 00:18:53   
4 2009-01-01 00:09:57           258 2009-01-01 00:11:58 2009-01-01 00:22:29   

  DeployedFromStation_Name  HourOfCall    IncidentGroup SpecialServiceType  \
0                  Clapham           0  Special Service                RTC   
1                 Edmonton           0             Fire                NaN   
2               Hillingdon           0             Fire                NaN   
3                 Holloway           0             Fire                NaN   
4                 Holloway           0             Fire                NaN   

   Easting_rounded  Northing_rounded color    d_long      d_lat         x_utm  \
0           528650            176850  blue -0.148916  51.475992 -16577.246612   
1           533750            194450   red -0.068836  51.632970  -7662.804053   
2           507750            182850   red -0.447902  51.534284 -49860.215886   
3           531050            185350   red -0.111220  51.551829 -12380.899219   
4           531050            185350   red -0.111220  51.551829 -12380.899219   

          y_utm    o_long      o_lat       o_x_utm       o_y_utm  \
0  6.705927e+06 -0.141051  51.465400 -15701.725496  6.704034e+06   
1  6.734032e+06 -0.069120  51.627217  -7694.403204  6.733000e+06   
2  6.716352e+06 -0.451701  51.531257 -50283.125311  6.715810e+06   
3  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   
4  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   

   dist_euclidian  
0     2085.533280  
1     1032.257024  
2      687.205480  
3     1872.413548  
4     1872.413548

# Il serait judicieux de calculer le temps de réaction des pompiers après un appel (càd combien de temps ils mettent pour prendre la route). Ceci servira notamment à comparer les casernes entre-elles, pour voir les plus efficaces selon ce critère:

In [29]:
liste_bis = []
for i in range(0, len(df)):
    liste_bis.append((df['TimeMobile'][i] - df['TimeOfCall'][i]).seconds)

In [30]:
Reaction_time = pd.DataFrame({'ReactionTime': liste_bis})

In [31]:
df = df.join(Reaction_time)

In [32]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

          TimeArrived  ResponseTime            TimeLeft        TimeReturned  \
0 2009-01-01 00:08:09           452 2009-01-01 00:12:17 2009-01-01 00:12:17   
1 2009-01-01 00:09:17           377 2009-01-01 00:16:33 2009-01-01 00:17:48   
2 2009-01-01 00:08:27           240 2009-01-01 00:20:38 2009-01-01 00:20:38   
3 2009-01-01 00:10:14           275                 NaT 2009-01-01 00:18:53   
4 2009-01-01 00:09:57           258 2009-01-01 00:11:58 2009-01-01 00:22:29   

  DeployedFromStation_Name  HourOfCall    IncidentGroup SpecialServiceType  \
0                  Clapham           0  Special Service                RTC   
1                 Edmonton           0             Fire                NaN   
2               Hillingdon           0             Fire                NaN   
3                 Holloway           0             Fire                NaN   
4                 Holloway           0             Fire                NaN   

   Easting_rounded  Northing_rounded color    d_long      d_lat         x_utm  \
0           528650            176850  blue -0.148916  51.475992 -16577.246612   
1           533750            194450   red -0.068836  51.632970  -7662.804053   
2           507750            182850   red -0.447902  51.534284 -49860.215886   
3           531050            185350   red -0.111220  51.551829 -12380.899219   
4           531050            185350   red -0.111220  51.551829 -12380.899219   

          y_utm    o_long      o_lat       o_x_utm       o_y_utm  \
0  6.705927e+06 -0.141051  51.465400 -15701.725496  6.704034e+06   
1  6.734032e+06 -0.069120  51.627217  -7694.403204  6.733000e+06   
2  6.716352e+06 -0.451701  51.531257 -50283.125311  6.715810e+06   
3  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   
4  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   

   dist_euclidian  ReactionTime  
0     2085.533280           363  
1     1032.257024           220  
2      687.205480           138  
3     1872.413548           167  
4     1872.413548           139

In [33]:
df.columns

Index(['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile',
       'TimeArrived', 'ResponseTime', 'TimeLeft', 'TimeReturned',
       'DeployedFromStation_Name', 'HourOfCall', 'IncidentGroup',
       'SpecialServiceType', 'Easting_rounded', 'Northing_rounded', 'color',
       'd_long', 'd_lat', 'x_utm', 'y_utm', 'o_long', 'o_lat', 'o_x_utm',
       'o_y_utm', 'dist_euclidian', 'ReactionTime'],
      dtype='object')

In [34]:
new_col_order = ['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile','ReactionTime',
       'TimeArrived', 'ResponseTime', 'TimeLeft', 'TimeReturned',
       'DeployedFromStation_Name', 'HourOfCall', 'IncidentGroup',
       'SpecialServiceType', 'Easting_rounded', 'Northing_rounded', 'color',
       'd_long', 'd_lat', 'x_utm', 'y_utm', 'o_long', 'o_lat', 'o_x_utm',
       'o_y_utm', 'dist_euclidian']

In [35]:
df = df.reindex(columns = new_col_order)

In [36]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

         TimeReturned DeployedFromStation_Name  HourOfCall    IncidentGroup  \
0 2009-01-01 00:12:17                  Clapham           0  Special Service   
1 2009-01-01 00:17:48                 Edmonton           0             Fire   
2 2009-01-01 00:20:38               Hillingdon           0             Fire   
3 2009-01-01 00:18:53                 Holloway           0             Fire   
4 2009-01-01 00:22:29                 Holloway           0             Fire   

  SpecialServiceType  Easting_rounded  Northing_rounded color    d_long  \
0                RTC           528650            176850  blue -0.148916   
1                NaN           533750            194450   red -0.068836   
2                NaN           507750            182850   red -0.447902   
3                NaN           531050            185350   red -0.111220   
4                NaN           531050            185350   red -0.111220   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
3  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
3  6.721272e+06     1872.413548  
4  6.721272e+06     1872.413548

# Nous pouvons aussi calculer le temps d'intervention de chaque opération:

In [37]:
liste_ter = []
for i in range(0, len(df)):
    liste_ter.append((df['TimeLeft'][i] - df['TimeArrived'][i]).seconds)

In [38]:
Inervention_duration = pd.DataFrame({"InterventionDuration": liste_ter})

In [39]:
df = df.join(Inervention_duration)

In [40]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

         TimeReturned DeployedFromStation_Name  HourOfCall    IncidentGroup  \
0 2009-01-01 00:12:17                  Clapham           0  Special Service   
1 2009-01-01 00:17:48                 Edmonton           0             Fire   
2 2009-01-01 00:20:38               Hillingdon           0             Fire   
3 2009-01-01 00:18:53                 Holloway           0             Fire   
4 2009-01-01 00:22:29                 Holloway           0             Fire   

   ... Northing_rounded  color    d_long      d_lat         x_utm  \
0  ...           176850   blue -0.148916  51.475992 -16577.246612   
1  ...           194450    red -0.068836  51.632970  -7662.804053   
2  ...           182850    red -0.447902  51.534284 -49860.215886   
3  ...           185350    red -0.111220  51.551829 -12380.899219   
4  ...           185350    red -0.111220  51.551829 -12380.899219   

          y_utm    o_long      o_lat       o_x_utm       o_y_utm  \
0  6.705927e+06 -0.141051  51.465400 -15701.725496  6.704034e+06   
1  6.734032e+06 -0.069120  51.627217  -7694.403204  6.733000e+06   
2  6.716352e+06 -0.451701  51.531257 -50283.125311  6.715810e+06   
3  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   
4  6.719493e+06 -0.116461  51.561766 -12964.379217  6.721272e+06   

   dist_euclidian  InterventionDuration  
0     2085.533280                 248.0  
1     1032.257024                 436.0  
2      687.205480                 731.0  
3     1872.413548                   NaN  
4     1872.413548                 121.0  

[5 rows x 26 columns]

# Mettons la colonne 'InterventionDuration' à l'index n°8

In [41]:
df.columns

Index(['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile',
       'ReactionTime', 'TimeArrived', 'ResponseTime', 'TimeLeft',
       'TimeReturned', 'DeployedFromStation_Name', 'HourOfCall',
       'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
       'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
       'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian',
       'InterventionDuration'],
      dtype='object')

In [42]:
new_col_order = ['IncidentNumber', 'TimeOfCall', 'TimeMobilised', 'TimeMobile',
       'ReactionTime', 'TimeArrived', 'ResponseTime', 'TimeLeft','InterventionDuration',
       'TimeReturned', 'DeployedFromStation_Name', 'HourOfCall',
       'IncidentGroup', 'SpecialServiceType', 'Easting_rounded',
       'Northing_rounded', 'color', 'd_long', 'd_lat', 'x_utm', 'y_utm',
       'o_long', 'o_lat', 'o_x_utm', 'o_y_utm', 'dist_euclidian']

In [43]:
df = df.reindex(columns = new_col_order)

In [44]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

   InterventionDuration        TimeReturned DeployedFromStation_Name  \
0                 248.0 2009-01-01 00:12:17                  Clapham   
1                 436.0 2009-01-01 00:17:48                 Edmonton   
2                 731.0 2009-01-01 00:20:38               Hillingdon   
3                   NaN 2009-01-01 00:18:53                 Holloway   
4                 121.0 2009-01-01 00:22:29                 Holloway   

   HourOfCall  ... Easting_rounded Northing_rounded  color    d_long  \
0           0  ...          528650           176850   blue -0.148916   
1           0  ...          533750           194450    red -0.068836   
2           0  ...          507750           182850    red -0.447902   
3           0  ...          531050           185350    red -0.111220   
4           0  ...          531050           185350    red -0.111220   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
3  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
3  6.721272e+06     1872.413548  
4  6.721272e+06     1872.413548  

[5 rows x 26 columns]

# La colonne 'TimeReturned' ne nous intéresse pas dans le cadre de notre problématique. Supprimons-la:

In [45]:
df.drop('TimeReturned', axis = 1, inplace = True)

In [46]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

   InterventionDuration DeployedFromStation_Name  HourOfCall    IncidentGroup  \
0                 248.0                  Clapham           0  Special Service   
1                 436.0                 Edmonton           0             Fire   
2                 731.0               Hillingdon           0             Fire   
3                   NaN                 Holloway           0             Fire   
4                 121.0                 Holloway           0             Fire   

  SpecialServiceType  Easting_rounded  Northing_rounded color    d_long  \
0                RTC           528650            176850  blue -0.148916   
1                NaN           533750            194450   red -0.068836   
2                NaN           507750            182850   red -0.447902   
3                NaN           531050            185350   red -0.111220   
4                NaN           531050            185350   red -0.111220   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
3  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
3  6.721272e+06     1872.413548  
4  6.721272e+06     1872.413548

# La colonne 'HourOfCall' ne nous intéresse pas vu que nous avons cette donnée dans 'TimeOfCall'. Supprimons-la: 

In [47]:
df.drop('HourOfCall', axis = 1, inplace = True)

In [48]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

   InterventionDuration DeployedFromStation_Name    IncidentGroup  \
0                 248.0                  Clapham  Special Service   
1                 436.0                 Edmonton             Fire   
2                 731.0               Hillingdon             Fire   
3                   NaN                 Holloway             Fire   
4                 121.0                 Holloway             Fire   

  SpecialServiceType  Easting_rounded  Northing_rounded color    d_long  \
0                RTC           528650            176850  blue -0.148916   
1                NaN           533750            194450   red -0.068836   
2                NaN           507750            182850   red -0.447902   
3                NaN           531050            185350   red -0.111220   
4                NaN           531050            185350   red -0.111220   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
3  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
3  6.721272e+06     1872.413548  
4  6.721272e+06     1872.413548

In [49]:
df.isnull().sum()

IncidentNumber                   0
TimeOfCall                       0
TimeMobilised                    0
TimeMobile                       0
ReactionTime                     0
TimeArrived                      0
ResponseTime                     0
TimeLeft                     42231
InterventionDuration         42231
DeployedFromStation_Name         0
IncidentGroup                    0
SpecialServiceType          914883
Easting_rounded                  0
Northing_rounded                 0
color                            0
d_long                           0
d_lat                            0
x_utm                            0
y_utm                            0
o_long                           0
o_lat                            0
o_x_utm                          0
o_y_utm                          0
dist_euclidian                   0
dtype: int64

# Nous remarquons qu'il y'a 3 colonnes avec des valeurs nulles, notamment "SpecialServiceType'. Voyons les valeurs uniques qu'elle possède: 

In [53]:
df['SpecialServiceType'].value_counts()

Flooding                           50048
Effecting entry/exit               49519
Lift Release                       49375
RTC                                46086
No action (not false alarm)        18119
Spills and Leaks (not RTC)         10940
Making Safe (not RTC)               8874
Assist other agencies               6375
Hazardous Materials incident        5682
Advice Only                         4632
Animal assistance incidents         4617
Other rescue/release of persons     4452
Evacuation (no fire)                3689
Other Transport incident            3345
Removal of objects from people      2178
Suicide/attempts                    2143
Medical Incident                    1782
Stand By                            1660
Medical Incident - Co-responder     1488
Rescue or evacuation from water      665
Water provision                      131
Name: SpecialServiceType, dtype: int64

# Il y'en a 27. Nous pouvons en déduire que les NaN veulent dirent qu'il n'y a pas de service spécial qui s'applique. Remplaçons ces NaN par la modalité 'None':

In [57]:
df['SpecialServiceType'].fillna('None', inplace = True)

In [58]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
3           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:08:26   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
3           167 2009-01-01 00:10:14           275                 NaT   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   

   InterventionDuration DeployedFromStation_Name    IncidentGroup  \
0                 248.0                  Clapham  Special Service   
1                 436.0                 Edmonton             Fire   
2                 731.0               Hillingdon             Fire   
3                   NaN                 Holloway             Fire   
4                 121.0                 Holloway             Fire   

  SpecialServiceType  Easting_rounded  Northing_rounded color    d_long  \
0                RTC           528650            176850  blue -0.148916   
1               None           533750            194450   red -0.068836   
2               None           507750            182850   red -0.447902   
3               None           531050            185350   red -0.111220   
4               None           531050            185350   red -0.111220   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
3  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
3  6.721272e+06     1872.413548  
4  6.721272e+06     1872.413548

In [59]:
df.isnull().sum()

IncidentNumber                  0
TimeOfCall                      0
TimeMobilised                   0
TimeMobile                      0
ReactionTime                    0
TimeArrived                     0
ResponseTime                    0
TimeLeft                    42231
InterventionDuration        42231
DeployedFromStation_Name        0
IncidentGroup                   0
SpecialServiceType              0
Easting_rounded                 0
Northing_rounded                0
color                           0
d_long                          0
d_lat                           0
x_utm                           0
y_utm                           0
o_long                          0
o_lat                           0
o_x_utm                         0
o_y_utm                         0
dist_euclidian                  0
dtype: int64

# Il reste deux colonnes avec des 42231 NaN chacune. Nous faisons le choix de les supprimer, vu que leur poids est relativement faible.

In [63]:
df.dropna(subset=['TimeLeft', 'InterventionDuration'], inplace = True)

In [66]:
df.isnull().sum()

IncidentNumber              0
TimeOfCall                  0
TimeMobilised               0
TimeMobile                  0
ReactionTime                0
TimeArrived                 0
ResponseTime                0
TimeLeft                    0
InterventionDuration        0
DeployedFromStation_Name    0
IncidentGroup               0
SpecialServiceType          0
Easting_rounded             0
Northing_rounded            0
color                       0
d_long                      0
d_lat                       0
x_utm                       0
y_utm                       0
o_long                      0
o_lat                       0
o_x_utm                     0
o_y_utm                     0
dist_euclidian              0
dtype: int64

In [67]:
df.head()

IncidentNumber          TimeOfCall       TimeMobilised          TimeMobile  \
0      235138081 2021-03-27 00:00:37 2009-01-01 00:02:27 2009-01-01 00:06:40   
1           2091 2021-03-27 00:03:00 2009-01-01 00:04:09 2009-01-01 00:06:40   
2           3091 2021-03-27 00:04:27 2009-01-01 00:04:57 2009-01-01 00:06:45   
4           5091 2021-03-27 00:05:39 2009-01-01 00:06:04 2009-01-01 00:07:58   
5           6091 2021-03-27 00:06:03 2009-01-01 00:06:30 2009-01-01 00:07:53   

   ReactionTime         TimeArrived  ResponseTime            TimeLeft  \
0           363 2009-01-01 00:08:09           452 2009-01-01 00:12:17   
1           220 2009-01-01 00:09:17           377 2009-01-01 00:16:33   
2           138 2009-01-01 00:08:27           240 2009-01-01 00:20:38   
4           139 2009-01-01 00:09:57           258 2009-01-01 00:11:58   
5           110 2009-01-01 00:09:22           199 2009-01-01 00:15:57   

   InterventionDuration DeployedFromStation_Name    IncidentGroup  \
0                 248.0                  Clapham  Special Service   
1                 436.0                 Edmonton             Fire   
2                 731.0               Hillingdon             Fire   
4                 121.0                 Holloway             Fire   
5                 395.0             Kentish Town      False Alarm   

  SpecialServiceType  Easting_rounded  Northing_rounded color    d_long  \
0                RTC           528650            176850  blue -0.148916   
1               None           533750            194450   red -0.068836   
2               None           507750            182850   red -0.447902   
4               None           531050            185350   red -0.111220   
5               None           529450            185250  cyan -0.134320   

       d_lat         x_utm         y_utm    o_long      o_lat       o_x_utm  \
0  51.475992 -16577.246612  6.705927e+06 -0.141051  51.465400 -15701.725496   
1  51.632970  -7662.804053  6.734032e+06 -0.069120  51.627217  -7694.403204   
2  51.534284 -49860.215886  6.716352e+06 -0.451701  51.531257 -50283.125311   
4  51.551829 -12380.899219  6.719493e+06 -0.116461  51.561766 -12964.379217   
5  51.551299 -14952.472965  6.719398e+06 -0.142053  51.552957 -15813.267626   

        o_y_utm  dist_euclidian  
0  6.704034e+06     2085.533280  
1  6.733000e+06     1032.257024  
2  6.715810e+06      687.205480  
4  6.721272e+06     1872.413548  
5  6.719694e+06      910.512670

# Notre df est prêt à l'analyse. 

In [68]:
df.to_csv(r'/home/smail/Bureau/data_with_coord_nettoye')